fix filesource for use with hostnames

benneti opened the pull request:
If I only set host and not address it is quite unexpected that a default address is used, therefore it makes more sense to try to make use of name resolution.
This shows another thing, it probably is unnecessary to have the option host and address, in the sense of ssh it would make more sense to just keep host and this can either be the hostname or the ip address (or is there a scenario I am missing where I need both?)

benneti commented:
If you agree that either the host or the address option is redundant tell me which you would like to keep and I'll change the PR accordingly

bordaigorl commented:
The change you are proposing makes sense.
The reason I was allowing both address and host is because one is used for the SSH client and the other for calling rsync. But it's more because of inexperience with ssh configuration I think...

benneti commented:
while doing this I realized a few more things and tried to improve them:

- a lot of code duplication between the init for ssh and rsync, changed that to call the init of ssh from rsync
- rsync wasn't respecting username setting as well as key setting, asuming that when I provide them here I want to use them I changed the code to follow that
- As I changed a lot I also changed the code to respect remote folder not corresponding to the remarkable directories, see https://github.com/bordaigorl/remy/issues/13
- Use shutil to use absolute path for rsync (and ssh)
- Lastly, try to give paramiko to find the key if none is provided by the conditions beforehand (worked for me, too)

I tested it with a simple ssh config and a few more complicated rsync setups and they seemed to be working.

If you want anything changed, let me know.

benneti commented:
Also is there a reason to not bulk download everything, i.e. this line
`    self._bulk_download(self._remote(), self._local(), includes=['*.metadata', '*.content', '*.pagedata'])
`
in combination with
`  def _bulk_download(self, fr, to, excludes=['*'], includes=[], delete=True):`.
As the remarkable should has only a few gb internal space anyway the rsync command should be fast even if one downloads everything from the folder?

If you have no preference I could add another option to sync everything (then the cache dir can function as a backup as well as a local source if there is no  connection)

bordaigorl commented:
I had a look, I'd be in favour of merging this but want to understand the rationale better.
There are three sets of proposed changes:
1. better ssh authentication
2. customisable remote root folders
3. [todo] add option to have the cache mirror remote folders

I am ok with 1, since I wrote that code I learned how to better deal with it, I may be importing some code from rmview actually. But this is a good first patch.

I am not sure I understand the use-case for 2 and I would not want behaviour that is counterintuitive.
If I understand correctly, you mirror all your tablet's content to a server.
You would like reMy to connect to that server to get the documents instead of the tablet's.
This however would, for example, imply that uploads are going to your server, so you'd need some service on the server that pushes them on the tablet (which sounds complicated and error prone without a proper protocol).
Am I misunderstanding this?

Change 3 seems something that could be useful as an option.
Proposal: `cache_mode` option with values
- `on_demand` default
- `full_mirror` download everything. I would still exclude the thumbnails and cache folders
I personally prefer the on demand version: I run my backups when I don't need to access the files. Running rsync for full copies is not cheap. The first load could take a long time as well.
My plan is to eventually incorporate the backup in the UI so that one can trigger it manually...

benneti commented:
Hi great to hear this.

I think you can rest assured as the changes for 2. do not influence the current behavior et al. i.e. if you do not set remote_documents or remote_templates in the remy.json file it will still use the known behavior and use the directories expected on the remarkable.
Yeah upload to the server would be really nice, but I do not think remy would be the best place to do this (or more precise if it works how I imagine the current functionality + custom remote path should be sufficient).
Actually rsync has an option "-u" to not replace newer files.
So if one feels like it, it would be possible to to push with remy to the server and let another systemd unit on the remarkable try to pull` rsync -u server local`.
Still this does not address the issue that one would have to restart the remarkable to actually see the documents...

I will implement 3. as you proposed. Should I add it to this PR, too?

bordaigorl commented:
Ok all clear. You can implement 3 here, it's fine.
Re 2: one just needs to make sure there are no expectations about things working if the defaults are changed.

benneti commented:
Hey, I am a bit puzzled right now, for some reason I think subprocess.run or something else tries to expand ~ in the call to rsync
`rsync: change_dir "/home/benneti//~/remarkable/templates" failed: No such file or directory (2)`.
while I have specified "~/remarkable/templates" as the appropriate remote directory, and if I use the rsync arguments from cmd it works without a problem.
I can live with using absolute paths for the remote directories but it still is weird for me, as a workaround I made a dummy version of expanduser for the remote, but this does not feel like a clean solution.

Furthermore I implemented a "full_mirror", as well as a "light_mirror" in addition to the standard "on_demand", see prefetchMetadata (I hope it is okay that I expanded this function, it seemed like the simplest approach to me.
Lastly `light_mirror` excludes the thumbnails folder but what cache folders should be excluded in addition?
The folder with the UUID in the name seems necessary because it contains the .rm files, does it not?

bordaigorl commented:
Thanks I'll have a look, probably in the weekend.
Have you tried using `$HOME` instead of `~`?

benneti commented:
Thank you, looking forward to the feedback. 

I did not try it in my configuration, the small helper _expanduser_remote uses echo $home and ssh to get the correct path and replaces the ~. I think like this the configuration behaves the most like one would exoect (in rsync command ~ works but $home woulad be interpilated by the shelk first such that i think using $home on the terminal is not possible).

benneti commented:
Is there anything else I could do or are you simply missing the time necessary to review?

bordaigorl commented:
Ah thanks for reminding me. Yes I am just very busy at the moment. Was planning some refactoring but will have to wait. I'll make a note to look at this next weekend.

benneti commented:
thank you!

bordaigorl commented:
Hi, PR looks mostly fine, but I don't like this "poor man" home expansion thing.
Have you tried adding the `--protect-args` option to rsync calls?

benneti commented:
I'll try to have a cleaner solution, else it probably is not a big problem to not be able to use "~" for the configuration of remote paths.
I still am curious where exactly and wny in the subprocess.run ~ leads to trouble (usual rsync calls work with ~).

benneti commented:
after some trial and error, I think ~ support on the remote is not worth the effort;
To be able to use this one would need to call subprocess.run with shell=True and when I do this I think the automatic argument handling breaks, i.e. it does not add appropriate " anymore.

benneti commented:
Feel free to merge if you are fine with the changes!

bordaigorl commented:
Thanks for the update. I am getting some errors when trying to connect with my config with the ssh method. I haven't tried to debug it yet, I'll see if I can find the problem.

benneti commented:
For me it still works, the source configuration I use to test it is:
``` json
    "ssh": {
      "name": "ssh",
      "type": "ssh",
      "host": "remarkable",
      "username": "root",
      "timeout": 10
    },
```
such that the known_hosts file is used to find the correct key.
If the error you get looks like it might not necessarily python you could try to delete the lines in .ssh/known_hosts and try whether this helps (I think it was not checked before and sometimes the remarkable changes its fingerprint leading to an error now that we check for compatibility of the fingerprint which I think is intended in the ssh protocol)

bordaigorl commented on remy/remarkable/filesource.py:
```diff
@@ -467,22 +483,20 @@ def fileSourceFromSSH(cls, name="SSH", address='10.11.99.1', username='root', pa
       'pkey': pkey,
       'timeout': timeout,
-      'look_for_keys': False
+      'look_for_keys': pkey is not None
```
Shouldn't this be `'look_for_keys': pkey is None`?

benneti replied:
fixed it

bordaigorl commented:
Hi, I don't know what I was doing wrong before, but I can confirm this works as expected.
The only thing that seems off is the line I commented on.
Once that's resolved I am ready to merge!
Next steps for me would be to make the connection code asynchronous to provide a better user experience.

benneti commented:
done, you are right; I guess my test configuration kept working because my key was already in my known_hosts file

benneti commented:
Also if you are working on the gui would you mind trying to implement bulk download (and download withou viewing) I actually tried to do it myself but I never worked with Qt or any gui stuff before and the effort seemed to high to grasp the code here...

bordaigorl commented:
> done, you are right; I guess my test configuration kept working because my key was already in my known_hosts file

Host keys are separate from the user key. The known hosts file stores the host key, while pkey stores the user key.

> Also if you are working on the gui would you mind trying to implement bulk download (and download withou viewing) I actually tried to do it myself but I never worked with Qt or any gui stuff before and the effort seemed to high to grasp the code here...

Those are all things on my todo list :) unfortunately I don't have much time to work on this at the moment...but as soon as I cook up something I'll push it, so watch the repo and you will get notified.
I may start from something quick like a context menu from the file browser so that viewing is optional.

Thanks for your contribution!
